velox/functions/prestosql/geospatial/GeometrySerde.h

```diff
@@ -17,17 +17,346 @@
+        numParts += polygon->getNumInteriorRing() + 1;
+      }
+    }
+
+    if (multiType) {
+      writer.write(
+          static_cast<uint8_t>(GeometrySerializationType::MULTI_POLYGON));
+    } else {
+      writer.write(static_cast<uint8_t>(GeometrySerializationType::POLYGON));
+    }
+
+    writer.write(static_cast<int32_t>(EsriShapeType::POLYGON));
+    writeEnvelope(geometry, writer);
+
+    writer.write(numParts);
+    writer.write(numPoints);
+
+    if (numParts == 0) {
+      return;
+    }
+
+    std::vector<int> partIndexes(numParts);
+    std::vector<bool> shellPart(numParts);
+
+    int currentPart = 0;
+    int currentPoint = 0;
+    for (int i = 0; i < numGeometries; i++) {
+      const geos::geom::Polygon* polygon =
+          dynamic_cast<const geos::geom::Polygon*>(geometry.getGeometryN(i));
+
+      partIndexes[currentPart] = currentPoint;
+      shellPart[currentPart] = true;
+      currentPart++;
+      currentPoint += polygon->getExteriorRing()->getNumPoints();
+
+      int holesCount = polygon->getNumInteriorRing();
+      for (int holeIndex = 0; holeIndex < holesCount; holeIndex++) {
+        partIndexes[currentPart] = currentPoint;
+        shellPart[currentPart] = false;
+        currentPart++;
+        currentPoint += polygon->getInteriorRingN(holeIndex)->getNumPoints();
+      }
+    }
+
+    for (int partIndex : partIndexes) {
+      writer.write(partIndex);
+    }
+
+    auto coordinates = geometry.getCoordinates();
+    canonicalizePolygonCoordinates(coordinates, partIndexes, shellPart);
+    writeCoordinates(coordinates, writer);
+  }
+
+  template <typename T>
+  static void writeGeometryCollection(
+      const geos::geom::Geometry& collection,
+      VarbinaryWriter<T>& output) {
+    output.write(
+        static_cast<uint8_t>(GeometrySerializationType::GEOMETRY_COLLECTION));
+
+    for (size_t geometryIndex = 0;
+         geometryIndex < collection.getNumGeometries();
+         ++geometryIndex) {
+      auto* geometry = collection.getGeometryN(geometryIndex);
+      // Use a temporary buffer to serialize the geometry and calculate its
+      // length
+      std::string tempBuffer;
+      VarbinaryWriter tempOutput(tempBuffer);
+
+      serialize(*geometry, tempOutput);
+
+      int32_t length = static_cast<int32_t>(tempBuffer.size());
+      output.write(length);
+      output.write(tempBuffer.data(), tempBuffer.size());
+    }
+  }
+};
+
+class GeometryDeserializer {
+ public:
+  /// Deserialize Velox's internal format to a geometry.  Do not call this
+  /// within GEOS_TRY macro: it will catch the exceptions that need to bubble
+  /// up.
+  static std::unique_ptr<geos::geom::Geometry> deserialize(
+      const StringView& geometryString) {
+    velox::common::InputByteStream inputStream(geometryString.data());
+    return deserialize(inputStream, geometryString.size());
+  }
+
+ private:
+  inline static auto geometryFactory_ = geos::geom::GeometryFactory::create();
```

Comment:
this would be used simultaneously among threads. will that usage be thread safe?

Reply:
But if we want to eliminate potential issues entirely, we can use a thread-local instance. That's my initial implementation.